pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,821 @@
+			}
+			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+		}
+		if len(stream.Status.DockerImageRepository) == 0 {
+			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
+			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
+			return "", e
+		}
+		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
+	}
+	return ref, nil
+}
+
+func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate, error) {
+
+	update := &buildUpdate{}
+
+	// Set the output Docker image reference.
+	ref, err := bc.resolveOutputDockerImageReference(build)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonInvalidOutputReference)
+		update.setMessage(buildapi.StatusMessageInvalidOutputRef)
+		return update, err
+	}
+
+	// Create the build pod spec
+	buildPod, err := bc.createPodSpec(build, ref)
+	if err != nil {
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPodSpec)
+		update.setMessage(buildapi.StatusMessageCannotCreateBuildPodSpec)
+		return update, err
+	}
+
+	glog.V(4).Infof("Pod %s/%s for build %s is about to be created", build.Namespace, buildPod.Name, buildDesc(build))
+	if _, err := bc.podClient.Pods(build.Namespace).Create(buildPod); err != nil {
+		if errors.IsAlreadyExists(err) {
+			bc.recorder.Eventf(build, kapi.EventTypeWarning, "FailedCreate", "Pod already exists: %s/%s", buildPod.Namespace, buildPod.Name)
+			glog.V(4).Infof("Build pod %s/%s for build %s already exists", build.Namespace, buildPod.Name, buildDesc(build))
+
+			// If the existing pod was created before this build, switch to the Error state.
+			existingPod, err := bc.podClient.Pods(build.Namespace).Get(buildPod.Name, metav1.GetOptions{})
+			if err == nil && existingPod.CreationTimestamp.Before(build.CreationTimestamp) {
+				update.setPhase(buildapi.BuildPhaseError)
+				update.setReason(buildapi.StatusReasonBuildPodExists)
+				update.setMessage(buildapi.StatusMessageBuildPodExists)
+				return update, nil
+			}
+			return nil, nil
+		}
+		// Log an event if the pod is not created (most likely due to quota denial).
+		bc.recorder.Eventf(build, kapi.EventTypeWarning, "FailedCreate", "Error creating: %v", err)
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPod)
+		update.setMessage(buildapi.StatusMessageCannotCreateBuildPod)
+		return update, fmt.Errorf("failed to create build pod: %v", err)
+	}
+	glog.V(4).Infof("Created pod %s/%s for build %s", build.Namespace, buildPod.Name, buildDesc(build))
+	update.setPodNameAnnotation(buildPod.Name)
+	update.setPhase(buildapi.BuildPhasePending)
+	update.setReason("")
+	update.setMessage("")
```

Comment:
If we're changing the phase to pending then it's not failed (according to the cache). If you mean that we should ensure that the build we're patching is the version that we know of in the cache, then an option is to specify the ResourceVersion in the patch. The patch would fail if the ResourceVersion has changed (per @deads2k). Would kind of defeat the purpose of switching to patch though.